### PR TITLE
Adding s3 client.

### PR DESCRIPTION
**Pull request by mpawliszyn**

### CODE REVIEW

File: client-s3/src/main/kotlin/app/cash/backfila/client/s3/record/StringStrategies.kt [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+class Utf8StringNewlineStrategy : RecordStrategy<String> {
+  override fun calculateNextRecordBytes(source: BufferedSource): Long = newlineByteCalculator(source)
+
+  override fun bytesToRecords(source: ByteString): List<String> = source.utf8().split("\n")
+}
+
+/**
+ * Optimized newline record byte calculator.
+ */
+fun newlineByteCalculator(source: BufferedSource): Long {
```
Comment (shellderp) [inline client-s3/src/main/kotlin/app/cash/backfila/client/s3/record/StringStrategies.kt:20]: should this be private?

File: client-s3/src/main/kotlin/app/cash/backfila/client/s3/record/StringStrategies.kt [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import okio.BufferedSource
+import okio.ByteString
+
+// This file contains some general strategies for handling String files for Backfills.
+
+/**
+ * Produces single string records separated by newlines.
+ */
+class Utf8StringNewlineStrategy : RecordStrategy<String> {
+  override fun calculateNextRecordBytes(source: BufferedSource): Long = newlineByteCalculator(source)
```
Comment (shellderp) [inline client-s3/src/main/kotlin/app/cash/backfila/client/s3/record/StringStrategies.kt:12]: what happens if you have a very long line in a file, I'm assuming the buffer here is not a full file

File: client-s3/src/test/kotlin/app/cash/backfila/client/s3/RecordStrategyBackfillTest.kt [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  @Inject lateinit var fakeS3: FakeS3Service
+
+  @BeforeEach
+  fun `loadFiles`() {
+    fakeS3.loadResourceDirectory("file-structure")
+  }
+
+  @Test
+  fun `backfilling with a record strategy that consumes no bytes fails`() {
+    with(SoftAssertions()) {
```
Comment (shellderp) [inline client-s3/src/test/kotlin/app/cash/backfila/client/s3/RecordStrategyBackfillTest.kt:35]: is there a benefit of soft assertions with only one assertion in the test? seems like a lot of unneeded boilerplate

File: client-s3/src/main/kotlin/app/cash/backfila/client/s3/internal/S3DatasourceBackend.kt [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+) : BackfillBackend {
+
+  /** Creates Backfill instances. Each backfill ID gets a new Backfill instance. */
+  private fun getBackfill(name: String): S3DatasourceBackfill<*, *>? {
+    val backfillClass = backfills[name]
+    return if (backfillClass != null) {
+      injector.getInstance(backfillClass.java) as S3DatasourceBackfill<*, *>
+    } else {
+      null
+    }
```
Comment (keeferrourke) [inline client-s3/src/main/kotlin/app/cash/backfila/client/s3/internal/S3DatasourceBackend.kt:36]: nit:

```suggestion
    return backfillClass?.let {    
      injector.getInstance(backfillClass.java) as S3DatasourceBackfill<*, *>
}
```

File: client-s3/src/main/kotlin/app/cash/backfila/client/s3/internal/S3DatasourceBackend.kt [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+  override fun create(backfillName: String): BackfillOperator? {
+    val backfill = getBackfill(backfillName)
+
+    if (backfill != null) {
+      @Suppress("UNCHECKED_CAST") // We don't know the types statically, so fake them.
+      return createS3DatasourceOperator(backfill as S3DatasourceBackfill<Any, Any>)
+    }
+
+    return null
```
Comment (keeferrourke) [inline client-s3/src/main/kotlin/app/cash/backfila/client/s3/internal/S3DatasourceBackend.kt:55]:
```suggestion
    return getBackfill(backfillName)?.let { backfill ->
      @Suppress("UNCHECKED_CAST") // We don't know the types statically, so fake them.
      createS3DatasourceOperator(backfill as S3DatasourceBackfill<Any, Any>)
    }
```

File: client-s3/src/main/kotlin/app/cash/backfila/client/s3/internal/S3DatasourceBackfillOperator.kt [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+      // This would mean we would need to support some kind of seek forward or seek back for record start.
+      // That or perhaps we only support it for single file?
+      // In any case, we are not implementing this now.
+      "Range is currently unsupported for S3 Backfils"
+    }
+
+    val pathPrefix = backfill.getPrefix(config)
+    val fileKeys = s3Service.listFiles(backfill.getBucket(config), pathPrefix)
+
+    // Sanity check that this backfill will actually process something
```
Comment (keeferrourke) [inline client-s3/src/main/kotlin/app/cash/backfila/client/s3/internal/S3DatasourceBackfillOperator.kt:42]:
```suggestion
    // Check that this backfill will actually process something
```

File: client-s3/src/main/kotlin/app/cash/backfila/client/s3/internal/S3DatasourceBackfillOperator.kt [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+            .build(),
+        )
+        .build()
+    }
+    return PrepareBackfillResponse.Builder()
+      .partitions(partitions)
+      .build()
+  }
+
+  override fun getNextBatchRange(request: GetNextBatchRangeRequest): GetNextBatchRangeResponse {
```
Comment (keeferrourke) [inline client-s3/src/main/kotlin/app/cash/backfila/client/s3/internal/S3DatasourceBackfillOperator.kt:73]: nit: this method is super long, is there a way to clean it up a bit by extracting some private methods?

File: client-s3/src/main/kotlin/app/cash/backfila/client/s3/internal/S3DatasourceBackfillOperator.kt [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    return GetNextBatchRangeResponse.Builder()
+      .batches(batches)
+      .build()
+  }
+
+  override fun runBatch(request: RunBatchRequest): RunBatchResponse {
+    val config = parametersOperator.constructBackfillConfig(request)
+    val pathPrefix = backfill.getPrefix(config.prepareConfig())
+    val batchRange = request.batch_range.decode()
+    require(batchRange.end != null) { "Batch was created without a range end." }
```
Comment (keeferrourke) [inline client-s3/src/main/kotlin/app/cash/backfila/client/s3/internal/S3DatasourceBackfillOperator.kt:158]:
```suggestion
    requireNotNull(batchRange.end) { "Batch was created without a range end." }
```

File: client-s3/src/main/kotlin/app/cash/backfila/client/s3/record/StringStrategies.kt (hunk 2 above)
Comment (keeferrourke) [reply client-s3/src/main/kotlin/app/cash/backfila/client/s3/record/StringStrategies.kt:12]: I think maybe you'd want a different strategy for splitting up the file in that case?

**Review by keeferrourke — APPROVED** [review]

File: client-s3/src/main/kotlin/app/cash/backfila/client/s3/record/StringStrategies.kt (hunk 2 above)
Comment (shellderp) [reply client-s3/src/main/kotlin/app/cash/backfila/client/s3/record/StringStrategies.kt:12]: then what's the line length limit supported, can we document it here?

File: client-s3/README.md [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+from this service. 
+
+The partition name will be the object name after the prefix. The range values reported are the byte seek 
+in the s3 object. The record counts are also reported in bytes rather than strict records. Scan size is also in bytes.
+
+In tests install the `FakeS3Module` and fill `FakeS3Service` with your test files. You have two options, 
+either add each file in code or point `FakeS3Service` to a resource path and it will load everything 
+under that path. Use `S3CdnModule` in real instances and provide an `AmazonS3` annotated with `@ForS3Backend`.
+
+Docs are hard, this may be out of date. Keep that in mind. The code is the source of truth.
```
[the remark this answers is not in the document]
Comment (shellderp) [reply client-s3/README.md:19]: seems a bit silly to write this lol. Should we put it in every doc file in this project?

File: client-s3/README.md [hunk 10]
```diff
@@ -0,0 +1,22 @@
+# client-s3
+
+Backfila client backend implementation to backfill all s3 objects in a bucket that match a prefix. 
+You specify a bucket via `getBucket`. You specify the prefix either statically by `staticPrefix` or by 
+overriding `getPrefix`. Each s3 object is a separate partition. You must define a record strategy to 
+divide up the s3 object into batches.
+* Object names must be less than 45 characters after the prefix.
```
Comment (shellderp) [inline client-s3/README.md:7]: where does this restriction come from?

File: client-s3/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (mpawliszyn) [reply client-s3/README.md:19]: It does seems silly but I get asked this all the time. This should encourage people to take a look at the code I would hope. I think we should add this everywhere yes.

I want to spend some time on docs, adding this kind of help to all our clients eventually.

File: client-s3/README.md (hunk 10 above)
Comment (shellderp) [reply client-s3/README.md:7]: then lets make the restriction 300 bytes? I don't think anyone uses nonascii filenames in s3

File: client-s3/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (shellderp) [reply client-s3/README.md:19]: in that case lets keep the docs simple and say something professional like "Refer to code for implementation details"

File: client-s3/README.md (hunk 10 above)
Comment (jaredwindover) [reply client-s3/README.md:7]: No strong feelings. I think 45 vs 75 was just an arbitrary buffer.

File: client-s3/src/main/kotlin/app/cash/backfila/client/s3/internal/S3DatasourceBackfillOperator.kt [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+) : BackfillOperator {
+
+  override fun name(): String = backfill.javaClass.toString()
+
+  override fun prepareBackfill(request: PrepareBackfillRequest): PrepareBackfillResponse {
+    val config = parametersOperator.constructBackfillConfig(request)
+    backfill.validate(config)
+
+    require(request.range?.start == null && request.range?.end == null) {
+      // We could think about supporting this later by making the range mean a byte seek into all S3 files.
```
Comment (shellderp) [inline client-s3/src/main/kotlin/app/cash/backfila/client/s3/internal/S3DatasourceBackfillOperator.kt:32]: doesn't make much sense to me to have a byte range into s3 files

File: client-s3/README.md (hunk 10 above)
Comment (mpawliszyn) [reply client-s3/README.md:7]: I'm going to leave this the way it is for now. I'm happy to change it later if it makes sense.

File: client-s3/src/main/kotlin/app/cash/backfila/client/s3/internal/S3DatasourceBackfillOperator.kt (hunk 11 above)
Comment (mpawliszyn) [reply client-s3/src/main/kotlin/app/cash/backfila/client/s3/internal/S3DatasourceBackfillOperator.kt:32]: If you completed part of the file and just want to continue? I'm open to thinking about it but it is pretty low as a priority.

File: client-s3/src/main/kotlin/app/cash/backfila/client/s3/internal/S3DatasourceBackfillOperator.kt (hunk 7 above)
Comment (mpawliszyn) [reply client-s3/src/main/kotlin/app/cash/backfila/client/s3/internal/S3DatasourceBackfillOperator.kt:73]: I'll do this after

File: client-s3/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (mpawliszyn) [reply client-s3/README.md:19]: It is a bit more than that though. I'm trying to nicely say, documentation is best effort, if you have questions read the f... code.

Let me try something else then:
"The code is the source of truth for this client. Keep that in mind. Always refer to the code for implementation details. If you notice a discrepancy please submit a pr."

Is that better? We can skip the PR sentance. I'm less sure about that.

[the remark this answers is not in the document]
Comment (shellderp) [reply client-s3/README.md:19]: yea either way

File: client-s3/src/main/kotlin/app/cash/backfila/client/s3/record/StringStrategies.kt (hunk 1 above)
Comment (shellderp) [reply client-s3/src/main/kotlin/app/cash/backfila/client/s3/record/StringStrategies.kt:20]: maybe.. I don't love that its global scope though

Comment (mpawliszyn) [reply client-s3/src/main/kotlin/app/cash/backfila/client/s3/record/StringStrategies.kt:20]: Sure I can do that StringStrategyTools?

File: client-s3/src/main/kotlin/app/cash/backfila/client/s3/record/StringStrategies.kt (hunk 2 above)
Comment (mpawliszyn) [reply client-s3/src/main/kotlin/app/cash/backfila/client/s3/record/StringStrategies.kt:12]: I'm pretty confident we don't actually want to limit this. The one thing we do check is that your strategy does always consume part of the stream so we are making progress.

File: client-s3/src/test/resources/file-structure/in-the-kitchen-with-mikepaw/dinner/main-butter-chicken [hunk 12]
```diff
@@ -0,0 +1,27 @@
+For Marinating the Chicken:
```
Comment (shellderp) [inline client-s3/src/test/resources/file-structure/in-the-kitchen-with-mikepaw/dinner/main-butter-chicken:1]: :chicken:

**Review by shellderp — APPROVED** [review]
solid

File: client-s3/src/test/resources/file-structure/in-the-kitchen-with-mikepaw/dinner/main-butter-chicken (hunk 12 above)
Comment (mpawliszyn) [reply client-s3/src/test/resources/file-structure/in-the-kitchen-with-mikepaw/dinner/main-butter-chicken:1]: Its a good butter chicken recipe.

Also I discovered relatively recently that butter chicken is basically tomato bisque...

**Pull request merged**